File: app/src/main/java/com/worldventures/dreamtrips/social/ui/infopages/presenter/EnrollRepPresenter.java
package com.worldventures.dreamtrips.social.ui.infopages.presenter;

import android.location.Location;

import com.worldventures.core.modules.infopages.StaticPageProvider;
import com.worldventures.dreamtrips.social.ui.infopages.util.PermissionLocationDelegate;
import com.worldventures.dreamtrips.social.ui.util.PermissionUIComponent;

import org.jetbrains.annotations.Nullable;

import javax.inject.Inject;

public class EnrollRepPresenter extends AuthorizedStaticInfoPresenter<EnrollRepPresenter.View> {

   @Inject StaticPageProvider staticPageProvider;
   @Inject PermissionLocationDelegate permissionLocationDelegate;

   public EnrollRepPresenter(String url) {
      super(url);
   }

   @Override
   public void takeView(EnrollRepPresenter.View view) {
      permissionLocationDelegate.setNeedRationalAction(view::showPermissionExplanationText);
      super.takeView(view);
   }

   @Override
   public void load() {
      permissionLocationDelegate.setLocationObtainedAction(location -> {
         updateUrlWithLocation(location);
         super.load();
      });
      permissionLocationDelegate.requestPermission(true, bindView());
   }

   @Override
   protected void reload() {
      permissionLocationDelegate.setLocationObtainedAction(location -> {
         updateUrlWithLocation(location);
         super.reload();
      });
      permissionLocationDelegate.requestPermission(true, bindView());
   }

   private void updateUrlWithLocation(@Nullable Location location) {
      if (location == null) {
         return;
      }
      url = staticPageProvider.getEnrollWithLocation(location.getLatitude(), location.getLongitude());
   }

   public void recheckPermissionAccepted(boolean recheckAccepted) {
      permissionLocationDelegate.recheckPermissionAccepted(recheckAccepted, bindView());
   }

   public interface View extends AuthorizedStaticInfoPresenter.View, PermissionUIComponent {

   }

}

File: app/src/main/java/com/worldventures/dreamtrips/social/ui/infopages/presenter/EnrollRepPresenter.kt
package com.worldventures.dreamtrips.social.ui.infopages.presenter

import android.location.Location

import com.worldventures.core.modules.infopages.StaticPageProvider
import com.worldventures.dreamtrips.social.ui.infopages.service.analytics.EnrolRepViewedAction
import com.worldventures.dreamtrips.social.ui.infopages.util.PermissionLocationDelegate
import com.worldventures.dreamtrips.social.ui.util.PermissionUIComponent

import javax.inject.Inject

class EnrollRepPresenter : AuthorizedStaticInfoPresenter<EnrollRepPresenter.View>() {

   @Inject lateinit var staticPageProvider: StaticPageProvider
   @Inject lateinit var permissionLocationDelegate: PermissionLocationDelegate

   override fun initUrl() = provider.enrollRepUrl ?: ""

   override fun takeView(view: EnrollRepPresenter.View) {
      permissionLocationDelegate.setNeedRationalAction(view::showPermissionExplanationText)
      super.takeView(view)
   }

   override fun load() {
      permissionLocationDelegate.setLocationObtainedAction {
         updateUrlWithLocation(it)
         super.load()
      }
      permissionLocationDelegate.requestPermission(true, bindView<Any>())
   }

   override fun reload() {
      permissionLocationDelegate.setLocationObtainedAction {
         updateUrlWithLocation(it)
         super.reload()
      }
      permissionLocationDelegate.requestPermission(true, bindView<Any>())
   }

   private fun updateUrlWithLocation(location: Location?) {
      location?.let { url = staticPageProvider.getEnrollRepWithLocation(it.latitude, it.longitude) }
   }

   fun recheckPermissionAccepted(recheckAccepted: Boolean) {
      permissionLocationDelegate.recheckPermissionAccepted(recheckAccepted, bindView<Any>())
   }

   fun sendPageDisplayedAnalyticsEvent(url: String?) {
      if (!url.isNullOrBlank() && this.url == url) {
         analyticsInteractor.analyticsActionPipe().send(EnrolRepViewedAction())
      }
   }

   interface View : AuthorizedStaticInfoPresenter.View, PermissionUIComponent

}
